### PR TITLE
Run multiple instance of scheduler on one JVM

### PR DESCRIPTION
**Pull request by xiangmo**
*Description of changes:*
Enable KCL V2 to run multiple instances of the Scheduler on a single JVM.
Multiple KCL schedulers should be able to run with different event source mappings of different customers in the same JVM.

Why?
Lambda event bridge application handles a high number of event source mappings of the customers. Running multiple KCL schedulers on the same instance and on the same JVM reduces the number of hosts required process the streams.

Solution:
There are static synchronized functions in KCL and they are not suitable for this scenario. They have to be modified to be synchronized per event source mapping and not per class.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/ShardSyncer.java [hunk 1]
```diff
@@ -49,12 +49,12 @@
...
 
 /**
- * Helper class to sync leases with shards of the Kinesis stream.
+ * Helper class per kcl scheduler instance to sync leases with shards of the Kinesis stream.
  * It will create new leases/activities when it discovers new Kinesis shards (bootstrap/resharding).
  * It deletes leases for shards that have been trimmed from Kinesis, or if we've completed processing it
  * and begun processing it's child shards.
  */
-@NoArgsConstructor(access = AccessLevel.PRIVATE)
+@NoArgsConstructor(access = AccessLevel.PUBLIC)
```
Comment (pfifer) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/ShardSyncer.java]: You shouldn't need to add this.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/ShardSyncTaskManager.java [hunk 2]
```diff
@@ -51,6 +51,28 @@
     private final ExecutorService executorService;
     @NonNull
     private final MetricsFactory metricsFactory;
+    @NonNull
+    private final ShardSyncer shardSyncer;
+
+    public ShardSyncTaskManager(ShardDetector shardDetector, LeaseRefresher leaseRefresher, InitialPositionInStreamExtended initialPositionInStream,
+                                boolean cleanupLeasesUponShardCompletion, boolean ignoreUnexpectedChildShards, long shardSyncIdleTimeMillis,
+                                ExecutorService executorService, MetricsFactory metricsFactory) {
+        this(shardDetector, leaseRefresher, initialPositionInStream, cleanupLeasesUponShardCompletion, ignoreUnexpectedChildShards, shardSyncIdleTimeMillis, executorService, metricsFactory, new ShardSyncer());
```
Comment (pfifer) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/ShardSyncTaskManager.java]: You should not depend on the `ShardSyncTaskManager` creating the `ShardSyncer`, instead handle the creation in `DynamoDBLeaseManagementFactory`.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/ShardSyncer.java (hunk 1 above)
Comment (xiangmo) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/ShardSyncer.java]: Removed

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/ShardSyncTaskManager.java (hunk 2 above)
Comment (xiangmo) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/ShardSyncTaskManager.java]: Modified.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/LeaseManagementConfig.java [hunk 3]
```diff
@@ -220,6 +220,15 @@ public LeaseManagementConfig metricsFactory(final MetricsFactory metricsFactory)
         }
     };
 
+    private ShardSyncer shardSyncer;
```
Comment (sahilpalvia) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/LeaseManagementConfig.java]: No need to create the getter

```java
private ShardSyncer shardSyncer = new ShardSyncer();
```

This should work.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/ShardSyncTaskManager.java [hunk 4]
```diff
@@ -51,6 +51,22 @@
     private final ExecutorService executorService;
     @NonNull
     private final MetricsFactory metricsFactory;
+    @NonNull
+    private final ShardSyncer shardSyncer;
+
+    public ShardSyncTaskManager(ShardDetector shardDetector, LeaseRefresher leaseRefresher, InitialPositionInStreamExtended initialPositionInStream,
```
Comment (sahilpalvia) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/ShardSyncTaskManager.java]: Could you change the order of MetricsFactory and ShardSyncer. We have maintained MetricsFactory as the last parameter of the constructors.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/ShardSyncer.java [hunk 5]
```diff
@@ -71,7 +68,7 @@
      * @throws KinesisClientLibIOException
      */
     // CHECKSTYLE:OFF CyclomaticComplexity
-    public static synchronized void checkAndCreateLeasesForNewShards(@NonNull final ShardDetector shardDetector,
+    public synchronized void checkAndCreateLeasesForNewShards(@NonNull final ShardDetector shardDetector,
```
Comment (sahilpalvia) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/ShardSyncer.java]: This seems like a breaking change. A better way would be to deprecate the existing method and creating a new non-static method.

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/coordinator/SchedulerTest.java [hunk 6]
```diff
@@ -259,6 +271,95 @@ public final void testInitializationFailureWithRetriesWithConfiguredMaxInitializ
         verify(shardDetector, times(maxInitializationAttempts)).listShards();
     }
 
+    /**
+     * Tests if two workers can concurrently sync shards.
+     * Runs the first worker and waits until it starts syncing shards. Then starts the second
+     * worker and checks if the second worker can start syncing shards.
+     */
+    @Test(timeout = 30000L)
+    public final void testWorkersCanSyncShardsInParallel() {
```
Comment (sahilpalvia) [inline amazon-kinesis-client/src/test/java/software/amazon/kinesis/coordinator/SchedulerTest.java]: Avoid multithreaded testing in unit tests. Testing a single instance should be fine for unit testing.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/ShardSyncTaskManager.java (hunk 4 above)
Comment (xiangmo) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/ShardSyncTaskManager.java]: You are correct. 
Modified

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/LeaseManagementConfig.java (hunk 3 above)
Comment (xiangmo) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/LeaseManagementConfig.java]: yes you are right

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/ShardSyncer.java (hunk 5 above)
Comment (xiangmo) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/ShardSyncer.java]: You are right, modified.

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/coordinator/SchedulerTest.java (hunk 6 above)
Comment (xiangmo) [reply amazon-kinesis-client/src/test/java/software/amazon/kinesis/coordinator/SchedulerTest.java]: OK, will remove this unit test.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/coordinator/Scheduler.java [hunk 7]
```diff
@@ -236,7 +239,8 @@ private void initialize() {
                 if (!skipShardSyncAtWorkerInitializationIfLeasesExist || leaseRefresher.isLeaseTableEmpty()) {
                     log.info("Syncing Kinesis shard info");
                     ShardSyncTask shardSyncTask = new ShardSyncTask(shardDetector, leaseRefresher, initialPosition,
-                            cleanupLeasesUponShardCompletion, ignoreUnexpetedChildShards, 0L, metricsFactory);
+                            cleanupLeasesUponShardCompletion, ignoreUnexpetedChildShards, 0L,
+                            hierarchichalShardSyncer, metricsFactory);
```
[the remark this answers is not in the document]
Comment (xiangmo) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/coordinator/Scheduler.java]: ye.. Fixed

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/ShardSyncTaskManager.java [hunk 8]
```diff
@@ -50,8 +50,24 @@
     @NonNull
     private final ExecutorService executorService;
     @NonNull
+    private final HierarchichalShardSyncer hierarchichalShardSyncer;
+    @NonNull
     private final MetricsFactory metricsFactory;
 
+    public ShardSyncTaskManager(ShardDetector shardDetector, LeaseRefresher leaseRefresher, InitialPositionInStreamExtended initialPositionInStream,
```
[the remark this answers is not in the document]
Comment (xiangmo) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/ShardSyncTaskManager.java]: Done

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/ShardSyncTaskManager.java [hunk 9]
```diff
@@ -81,7 +97,7 @@ private synchronized boolean checkAndSubmitNextTask() {
                                     initialPositionInStream,
                                     cleanupLeasesUponShardCompletion,
                                     ignoreUnexpectedChildShards,
-                                    shardSyncIdleTimeMillis,
+                                    shardSyncIdleTimeMillis, hierarchichalShardSyncer,
```
[the remark this answers is not in the document]
Comment (xiangmo) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/ShardSyncTaskManager.java]: Fixed

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ConsumerStatesTest.java [hunk 10]
```diff
@@ -109,7 +110,8 @@ public void setup() {
                 taskBackoffTimeMillis, skipShardSyncAtWorkerInitializationIfLeasesExist,
                 listShardsBackoffTimeInMillis, maxListShardsRetryAttempts,
                 shouldCallProcessRecordsEvenForEmptyRecordList, idleTimeInMillis, INITIAL_POSITION_IN_STREAM,
-                cleanupLeasesOfCompletedShards, ignoreUnexpectedChildShards, shardDetector, metricsFactory, new AggregatorUtil());
+                cleanupLeasesOfCompletedShards, ignoreUnexpectedChildShards, shardDetector, new AggregatorUtil(),
+                new HierarchichalShardSyncer(), metricsFactory);
```
[the remark this answers is not in the document]
Comment (xiangmo) [reply amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ConsumerStatesTest.java]: before this change, you used static methods in ShardSyncer.class and you never mock it. That's why I think we need to create a new instance.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseManagementFactory.java [hunk 11]
```diff
@@ -150,6 +157,54 @@ public DynamoDBLeaseManagementFactory(final KinesisAsyncClient kinesisClient, fi
...
+     * @param listShardsBackoffTimeMillis
+     * @param maxListShardsRetryAttempts
+     * @param maxCacheMissesBeforeReload
+     * @param listShardsCacheAllowedAgeInSeconds
+     * @param cacheMissWarningModulus
+     * @param initialLeaseTableReadCapacity
+     * @param initialLeaseTableWriteCapacity
+     * @param hierarchichalShardSyncer
+     */
+    public DynamoDBLeaseManagementFactory(final KinesisAsyncClient kinesisClient, final String streamName,
```
[the remark this answers is not in the document]
Comment (xiangmo) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseManagementFactory.java]: Done

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShutdownTaskTest.java [hunk 12]
```diff
@@ -54,6 +55,7 @@
             InitialPositionInStreamExtended.newInitialPosition(InitialPositionInStream.TRIM_HORIZON);
     private static final ShutdownReason TERMINATE_SHUTDOWN_REASON = ShutdownReason.SHARD_END;
     private static final MetricsFactory NULL_METRICS_FACTORY = new NullMetricsFactory();
+    private static final HierarchichalShardSyncer SHARD_SYNCER = new HierarchichalShardSyncer();
```
[the remark this answers is not in the document]
Comment (xiangmo) [reply amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShutdownTaskTest.java]: DOne

File: amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ConsumerStatesTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (xiangmo) [reply amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ConsumerStatesTest.java]: Changed to use mock

**Comment by pfifer** [comment]
Thanks for the change, and sorry about the delay.

One more thing can you please pull resolve the conflicts, and update the PR.

**Pull request merged**